File: 0948-sort-an-array/0948-sort-an-array.cpp
//MergeSort,  T.C -> O(nlogn),  S.C -> O(n)
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        mergeSort(nums, 0, nums.size()-1);
        return nums;
    }
    void mergeSort(vector<int>&nums, int start, int end){
        if(start < end){
            int mid = start + (end - start)/2;
            mergeSort(nums, start, mid);
            mergeSort(nums, mid+1, end);
            merge(nums, start, mid, end);
        }
    }
    void merge(vector<int> &nums, int low, int mid, int high){
        if(low >= high) return;
        int l = low, r = mid+1, k = 0, sz = high - low + 1;
        vector<int> sorted(sz, 0);
        while(l <= mid && r <= high) sorted[k++] = nums[l] < nums[r] ? nums[l++] : nums[r++];
        while(l <= mid) sorted[k++] = nums[l++];
        while(r <= high) sorted[k++] = nums[r++];
        for(k=0; k<sz; k++) nums[k+low] = sorted[k];
    }
};

File: 0948-sort-an-array/0948-sort-an-array.java
class Solution {
    public int[] sortArray(int[] nums) {
        mergeSort(nums, 0, nums.length-1);
        return nums;
    }
    void mergeSort(int[] nums, int l, int r){
        if(l < r){
            int m = (l+r)/2;
            mergeSort(nums, l, m);
            mergeSort(nums, m+1, r);
            merge(nums, l, m, r);
        }
    }
    void merge(int[] nums, int l, int m, int r){
        if(l >= r) return;
        int s = l, e = m+1, n = r-l+1, k=0;
        int[] sortedArray = new int[n];
        while(s<=m && e<=r) sortedArray[k++] = nums[s]<nums[e] ? nums[s++] : nums[e++];
        while(s <= m) sortedArray[k++] = nums[s++];
        while(e <= r) sortedArray[k++] = nums[e++];
        for(k=0; k<n; k++) nums[l+k] = sortedArray[k];
    }
}